### ISSUE TITLE
Improve autotemplate algorithm

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#97]
`chezmoi` uses [`text/template`](https://godoc.org/text/template) as its templating language. For an example of `text/template` in action, see https://play.golang.org/p/lzBUhUAEWXk.

[`text/template.Template.Execute`](https://godoc.org/text/template#Template.Execute) takes as input a template and some `data` and outputs the template with `data` substituted. For `chezmoi`'s autotemplate function, we want to reverse this process. Specifically:

> Given some `output string` and some `data map[string]string`, determine the optimal `templateStr string` such that executing the template returned by `template.New("").Parse(templateStr)` with `data` results in `output`.

Requirements:
 * Direct substitution should work, e.g. given `output=user@home.org` and `data := map[string]string{"email": "user@home.org"}` the resulting template should be `output={{ .email }}`.
 * Longest matches should win, e.g. given `name=John Smith` and `data := map[string]string{"firstName":"John", "fullName":"John Smith"}` the output should be `name={{  .fullName }}`, not `name={{ .firstName}} Smith`.
* If multiple template variables have the same value, then the alphabetically first should win, e.g. given `name=foo` and `data := map[string]string{"a": "foo", "b": "foo"}` the output should be `name={{ .a }}`, not `name={{ .b }}`.
* Substitution should not occur within words, e.g. given `os=darwinian evolution` and input `"data := map[string]string{"os": "darwin"}` the output should be `os=darwinian evolution`, not `os={{ .os }}ian evolution`.
* Substitution should be flexible on word boundaries, e.g. given output `/home/user` and `data := map[string]string{"h": "/home/user", "u": "user"}` the output should be `{{ .h }}` and not `/home/{{ .u }}`.

There's an [existing algorithm](https://github.com/twpayne/chezmoi/blob/afd6994546b541b2c63acda0c6f009e1cf42ac64/lib/chezmoi/autotemplate.go#L44-L77) and [existing tests](https://github.com/twpayne/chezmoi/blob/afd6994546b541b2c63acda0c6f009e1cf42ac64/lib/chezmoi/autotemplate_test.go#L5-L122), but the algorithm is slow (maybe O(N^2) at best) and the tests are incomplete.

Points for:
* Adding tests that demonstrate where the current algorithm fails.
* Writing code that passes these tests.
* Achieving better time and space complexity.
* Achieving better runtime.
* Supporting Unicode (`[]rune`), not strings (`[]byte`).

**Comment by twpayne** [comment]
This is a nice-to-have, and not needed, especially after #393. Please re-open if you have some good ideas.
